api/test/integration/indicator-record/indicator-records.service.spec.ts

```diff
@@ -0,0 +1,364 @@
+      };
+
+      //ACT
+      await indicatorRecordService.createIndicatorRecordsBySourcingRecords(
+        sourcingData,
+        providedCoefficients,
+      );
+
+      //ASSERT
+      const allIndicators = await indicatorRecordRepository.find();
+      expect(allIndicators.length).toEqual(4);
+
+      //Value is provided coeff * tonnage, and scaler is null, because it's based on production coefficient, and it's not provided
+      await checkCreatedIndicatorRecord(
+        INDICATOR_TYPES.DEFORESTATION,
+        indicatorPreconditions.deforestation,
+        sourcingData,
+        350,
+        null,
+      );
+      await checkCreatedIndicatorRecord(
+        INDICATOR_TYPES.BIODIVERSITY_LOSS,
+        indicatorPreconditions.biodiversityLoss,
+        sourcingData,
+        100,
+        null,
+      );
+      await checkCreatedIndicatorRecord(
+        INDICATOR_TYPES.CARBON_EMISSIONS,
+        indicatorPreconditions.carbonEmissions,
+        sourcingData,
+        400,
+        null,
+      );
+      await checkCreatedIndicatorRecord(
+        INDICATOR_TYPES.UNSUSTAINABLE_WATER_USE,
+        indicatorPreconditions.waterRisk,
+        sourcingData,
+        200,
+        null,
+      );
+    });
+
+    it("When creating indicator record with no provided coefficients, and there's no H3 data for the given material, it should throw an error it should create corresponding CANCELLED Sourcing Locations, with the appropiate Sourcing Records and Indicator Records", async () => {
+      //ARRANGE
+      const sourcingData = {
+        sourcingRecordId: UUIDv4(),
+        geoRegionId: UUIDv4(),
+        materialId: UUIDv4(),
+        tonnage: 10000,
+      };
+
+      jest.spyOn(materialsToH3sService, 'findOne').mockResolvedValue(undefined);
+
+      const testStatement = async (): Promise<any> => {
+        await indicatorRecordService.createIndicatorRecordsBySourcingRecords(
+          sourcingData,
+        );
+      };
+
+      //ACT/ASSERT
+      await expect(testStatement).rejects.toThrow(MissingH3DataError);
+      await expect(testStatement).rejects.toThrow(
+        `No H3 Data required for calculate Impact for Material with ID: ${sourcingData.materialId}`,
+      );
+    });
+
+    it('When creating indicators without provided coefficients and the material has H3 data, it should create the indicator records properly', async () => {
+      //ARRANGE
+      const indicatorPreconditions = await createPreconditions();
+
+      const sourcingData = {
+        sourcingRecordId: indicatorPreconditions.sourcingRecord2.id,
+        tonnage: indicatorPreconditions.sourcingRecord2.tonnage,
+        geoRegionId: indicatorPreconditions.sourcingLocation2.geoRegionId,
+        materialId: indicatorPreconditions.sourcingLocation2.materialId,
+      };
+
+      // The original function only expects for one MaterialTo3 data, the values don't matter
+      jest.spyOn(materialsToH3sService, 'findOne').mockResolvedValue({
```

Comment:
One note on the mocking side:

If you need to use the same mock for all the test cases, maybe it is better to use NestJs feature:

```
.overrideProvided(ClassToOverride).useClass(YourMockClass) 
```
or 
```
.overrideProvided(ClassToOverride).useValue(yourMockObject) 
```

Also, if you do need to mock different stuff, it is safer to use
```
jest.clearAllMocks()
```
In your afterEach() hook to be sure all mocks are clear within tests cases and avoid unexpected behaviours

Reply:
I see what you mean, but considering it's only mocked two times, I thought it maybe was worth it, when only mocking the result was needed.
It is true though that as it is right, is not properly the mocking result is not properly cleaned up. When using a mock function/value on a spied function, the mock will remain there until .restore() or jset-resotreAllMocks() is called. I should have used mockResolvedValue**Once**, that way once the mock has been executed once, the original behaviour will be used again. I'll change this for better, consistent behaviour.